### ISSUE TITLE
New rule: No promise inside if condition

### ISSUE THREAD

**Issue by joshuaavalon** [eslint/eslint#11525]
**Please describe what the rule should do:**

The rule should warn about putting promise in a if condition because it always resolve to true.

In most of the cases, we want to use the result of the promise instead.

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```javascript
async function isValid() {
  return false;
}

if (isValid()) { // This is most likely an error because promise always resolves to true
  // Do work
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is a common mistake instead of code style. I think it should be included in ESLint to avoid this mistake.

**Are you willing to submit a pull request to implement this rule?**

No, I have do not any experience with ESLint code base.

**Comment by platinumazure** [comment]
Hi @joshuaavalon, thanks for the issue!

I like the idea, but I don't think this is something we could do reliably unless the function in question is in the same file (so we can statically determine that it is async). There would be no way to tell for imported functions, unless you're using Typescript or similar.

I'm on the fence about including a rule in core for the limited case. If the team doesn't accept this proposal, maybe an issue could be raised in [eslint-plugin-promise](https://github.com/xjamundx/eslint-plugin-promise) to support this case.

**Comment by joshuaavalon** [comment]
@platinumazure How about `async function` instead of `Promise`? `async function` always return a `Promise`.

**Comment by platinumazure** [comment]
@joshuaavalon Yes, if the function can be identified as async via static analysis it could be flagged easily if it is invoked in an if condition without an await. However, if the function is imported from another file, there's no way to tell.

**Comment by joshuaavalon** [comment]
The limitation of JavaScript prevents this rule to be useful.

I am closing this issue and created a proposal at typescript-eslint/typescript-eslint#365.
